### PR TITLE
= http: #1014 Fix handling of empty authority in Uris

### LINKED ISSUE #1014: Uri Authority instances with empty host but non-empty port or user-info not handled correctly
Backport of https://github.com/akka/akka/issues/16953.
Fix in https://github.com/akka/akka/pull/16954.

Easy to backport.

### PR DESCRIPTION
**Pull request by wjur**
It is a backport of https://github.com/akka/akka/pull/16954 and it fixes #1014. 
Additionally, this PR fixes a couple of edge-cases in rendering slashes succeeding the authority.

### CODE REVIEW

File: spray-http/src/test/scala/spray/http/UriSpec.scala [hunk 1]
```diff
@@ -358,8 +366,12 @@ class UriSpec extends Specification {
       normalize("Http://Localhost") === "http://localhost"
       normalize("hTtP://localHost") === "http://localhost"
       normalize("https://:443") === "https://"
+      normalize("https://:444") === "https://:444"
+      normalize("http://:80/foo") === "http:///foo"
+      normalize("http://:8080/foo") === "http://:8080/foo"
       normalize("ftp://example.com:21") === "ftp://example.com"
-      normalize("example.com:21") === "example.com:21"
+      normalize("example.com:21") === "example.com:21" // example.com is parsed as the SCHEME (which is correct)
```
Comment (wjur) [inline spray-http/src/test/scala/spray/http/UriSpec.scala]: In Akka, this test expects "example.com:21" to be normalized to example.com://21 (and the corresponding PR changes this line accordingly). I guess this case is analogous to the "g:h"-case so no slashes should appear here as well.

**Comment by wjur** [comment]
Ok, I see the build failed on OpenJdk. I humbly admit that I have checked if it's working on Oracle JVM. I need to investigate this...

**Comment by sirthias** [comment]
@wjur Don't worry about the failing test. Some of our tests are a bit shaky.
I'll look into your changes as soon as my time permits. Thanks a lot again!!

File: spray-http/src/main/scala/spray/http/Uri.scala [hunk 2]
```diff
@@ -158,10 +158,9 @@ sealed abstract case class Uri(scheme: String, authority: Authority, path: Path,
    */
   def renderWithoutFragment[R <: Rendering](r: R, charset: Charset): r.type = {
     if (isAbsolute) r ~~ scheme ~~ ':'
-    authority.render(r, scheme, charset)
+    authority.render(r, scheme, path.isEmpty || path.startsWithSlash, charset)
```
Comment (sirthias) [inline spray-http/src/main/scala/spray/http/Uri.scala]: Could we maybe simply pass the `path` to the render method and have the `path.isEmpty || path.startsWithSlash` condition be evaluated there?
This would have the benefit of keeping all the logic required for rendering inside of the `render` method, rather than moving some of it outside into argument construction.

**Comment by sirthias** [comment]
This looks great, thank you @wjur!
Maybe one more small tweak and we can merge.

**Comment by wjur** [comment]
I have updated the PR according to your comment.

File: spray-http/src/main/scala/spray/http/Uri.scala [hunk 3]
```diff
@@ -330,18 +329,19 @@ object Uri {
...
 
   case class Authority(host: Host, port: Int = 0, userinfo: String = "") extends ToStringRenderable {
-    def isEmpty = host.isEmpty
-    def render[R <: Rendering](r: R): r.type = render(r, "", UTF8)
-    def render[R <: Rendering](r: R, scheme: String, charset: Charset): r.type =
-      if (isEmpty) r else {
+    def isEmpty = equals(Authority.Empty)
+    def nonEmpty = !isEmpty
+    def render[R <: Rendering](r: R): r.type = render(r, "", Path.Empty, UTF8)
+    def render[R <: Rendering](r: R, scheme: String, charset: Charset): r.type = render(r, scheme, Path.Empty, charset)
```
Comment (wjur) [inline spray-http/src/main/scala/spray/http/Uri.scala]: Previously I deleted this render method (the one with 3 arguments: R, String and Charset) but now I have reverted it (for compatibility). Maybe someone is using it outside spray.

Comment (sirthias) [inline spray-http/src/main/scala/spray/http/Uri.scala]: Yes, thanks. There is no reason to change anything about existing API.

**Comment by sirthias** [comment]
Ok, thanks. This LGTM.
@jrudolph ?

**Comment by jrudolph** [comment]
Yep, LGTM. Thanks, @wjur.

**Comment by sirthias** [comment]
Ok, great.
Now we'd need a PR in Akka, against branch `release-2.3-dev` containing the fix to the fix.
You said you also would be able to do this, @wjur ?

**Comment by wjur** [comment]
Thank you for the merge, guys.
Yes, I will prepare the fix for Akka (or at least try to prepare).

**Comment by wjur** [comment]
@sirthias I have prepared the PR: https://github.com/akka/akka/pull/17052
